record/internal.go

```diff
@@ -0,0 +1,371 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package record
+
+import (
+	"errors"
+	"hash"
+	"hash/crc32"
+	"math"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/prometheus/tsdb/chunkenc"
+	"github.com/prometheus/tsdb/chunks"
+	"github.com/prometheus/tsdb/fileutil"
+	"github.com/prometheus/tsdb/labels"
+)
+
+var (
+	// ErrOutOfOrderSample is returned if an appended sample has a
+	// timestamp smaller than the most recent sample.
+	ErrOutOfOrderSample = errors.New("out of order sample")
+
+	// ErrNotFound is returned if a looked up resource was not found.
+	ErrNotFound = errors.New("not found")
+
+	// ErrAmendSample is returned if an appended sample has the same timestamp
+	// as the most recent sample but a different value.
+	ErrAmendSample = errors.New("amending sample")
+)
+
+// The table gets initialized with sync.Once but may still cause a race
+// with any other use of the crc32 package anywhere. Thus we initialize it
+// before.
+var castagnoliTable *crc32.Table
+
+func init() {
+	castagnoliTable = crc32.MakeTable(crc32.Castagnoli)
+}
+
+// NewCRC32 initializes a CRC32 hash with a preconfigured polynomial, so the
+// polynomial may be easily changed in one location at a later time, if necessary.
+func NewCRC32() hash.Hash32 {
+	return crc32.New(castagnoliTable)
+}
+
+type sample struct {
+	t int64
+	v float64
+}
+
+func (s sample) T() int64 {
+	return s.t
+}
+
+func (s sample) V() float64 {
+	return s.v
+}
+
+// SizeReader returns the size of the object in bytes.
+type SizeReader interface {
+	// Size returns the size in bytes.
+	Size() int64
+}
+
+// RefSeries is the series labels with the series ID.
+type RefSeries struct {
+	Ref    uint64
+	Labels labels.Labels
+}
+
+// RefSample is a timestamp/value pair associated with a reference to a series.
+type RefSample struct {
+	Ref    uint64
+	T      int64
+	V      float64
+	Series *MemSeries
+}
+
+// MemSeries is the in-memory representation of a series. None of its methods
+// are goroutine safe and it is the caller's responsibility to lock it.
+type MemSeries struct {
```

Comment:
This only seems to be used by the root package. Is there a reason it had to move that I am missing?

Reply:
It's also used by the `RefSample` type just above it in this same package. I thought it made sense to keep them both in the same package.

Reply:
Hmmm... How important do you think it is to have a pointer to `MemSeries` on `RefSample`? Having the two tied together really only makes sense in the head, whereas we now use RefSample in a number of places.